rp2040-hal/src/multicore.rs

```diff
@@ -155,12 +155,15 @@ impl<'p> Core<'p> {
             extern "C" fn core1_startup<F: FnOnce() -> bad::Never>(
                 _: u64,
                 _: u64,
-                entry: &mut ManuallyDrop<F>,
+                entry: *mut ManuallyDrop<F>,
                 stack_bottom: *mut usize,
             ) -> ! {
                 core1_setup(stack_bottom);
 
-                let entry = unsafe { ManuallyDrop::take(entry) };
+                let entry = unsafe { ManuallyDrop::take(&mut *entry) };
```

Comment:
I'd prefer to explicitly read the pointer value, instead of just dereferencing it.
`&mut *entry` still points to memory on the core0 stack, and `ManuallyDrop::take` is not guaranteed to copy the contents somewhere else.

```suggestion
                let entry = unsafe { ManuallyDrop::take(&mut core::ptr::read(entry)) };
```

Reply:
I think it's sound either with or without the `ptr::read()`. `ptr::read()` can still be optimized away, it's logically the same as a regular dereference. Only `ptr::read_volatile()` is guaranteed to not be optimized away.

Reply:
> But if the value is used, the read must happen before the fence. That's guaranteed by the ptr::read.

it is also guaranteed by the dereference. It is also a memory read (from the pointer to a local variable).

> we can't be sure that the original location doesn't alias. In fact we know that it does.

The `*mut` pointer is borrowed from the original `entry` in core0 stack, so we are allowed to access it even with `&mut`. Only when core0 touches `entry` again, the `*mut` is invalidated (see Stacked Borrows / Tree Borrows). We know that doesn't happen because core0 doesn't do so until we tell it it's fine to do. And when we do, the `&mut` already no longer exists.

Reply:
`ManuallyDrop::take`  does the `ptr::read`. That's actually the only thing it does.

I am wondering why don't we pass in the `ManuallyDrop<F>` directly rather than an `*mut` to it?
As a far as I can see from a little godbolt experiment, the only difference is that when using `ManuallyDrop::into_inner` it does not seem able to lower into a memcpy while this is what is done when using `ManuallyDrop::take`.
IMHO this is really an issue only if F is large, and F must not be large.

Reply:
> F must not be large

Why?

Reply:
Because a large object on a stack that has no enforced limit is unsound and likely to (unexpectedly?) overflow.

Reply:
My Extremely limited understanding is that &mut has a lifetime attached while *mut does not. And that might not even be true 😅

Reply:
> I am wondering why don't we pass in the `ManuallyDrop<F>` directly rather than an `*mut` to it?

We place the parameters on the stack manually, and need to follow the AAPCS32 calling convention standards. For the pointer/reference, that's easy. But a `ManuallyDrop<F>` has the same memory layout as `F`, ie. it depends on what `F` is. That _could_ become difficult. For example, what happens if `F` contains a field with unusual alignment requirements?